subcommands/waves/init.go

```diff
@@ -0,0 +1,132 @@
+package waves
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/docker/go/canonical/json"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"github.com/foundriesio/fioctl/client"
+	"github.com/foundriesio/fioctl/subcommands"
+	tuf "github.com/theupdateframework/notary/tuf/data"
+)
+
+func init() {
+	initCmd := &cobra.Command{
+		Use:   "init <wave> <version> [<tag>]",
+		Short: "Create a new wave for a given targets version.",
+		Long: `Create a new wave for a given targets version.",
+Optionally, provide a new tag to use for these targets ('master' by default).
+In any case, original tags of these targets set before creating a wave are ignored.
+
+This command only initializes a wave, but does not provision its updates to devices.
+Use a "fioctl wave rollout <wave> <group>" to trigger updates of this wave to one group.
+Use a "fioctl wave promote <wave>" to update all devices (make it globally available).`,
+		Run:  doInitWave,
+		Args: cobra.RangeArgs(2, 3),
+	}
+	cmd.AddCommand(initCmd)
+	initCmd.Flags().IntP("expires-days", "e", 0, `Role expiration in days; default 90.
+The same expiration will be used when a wave is promoted.`)
+	initCmd.Flags().StringP("expires-at", "E", "", `Role expiration date and time in RFC 3339 format.
+The same expiration will be used when a wave is promoted.
+When set this value overrides an 'expires-at' argument.
+Example: 2020-01-01T00:00:00Z`)
+	initCmd.Flags().BoolP("dry-run", "d", false, "Don't create a wave, print it to standard output.")
+	// TODO: Add an optional offline signing arguments
+}
+
+func doInitWave(cmd *cobra.Command, args []string) {
+	factory := viper.GetString("factory")
+	name, version, tag := args[0], args[1], "master"
+	if len(args) > 2 {
+		tag = args[2]
+	}
+	expires := readExpiration(cmd)
+	dryRun, _ := cmd.Flags().GetBool("dry-run")
+	logrus.Debugf("Creating a wave %s for factory %s targets version %s and new tag %s expires %s",
+		name, factory, version, tag, expires.Format(time.RFC3339))
+
+	new_targets, err := api.TargetsList(factory, version)
+	subcommands.DieNotNil(err)
+	if len(new_targets) == 0 {
+		subcommands.DieNotNil(fmt.Errorf("No new targets of a given version"))
+	}
+
+	// TODO: Fetch current promoted targets for a given tag.
+	// There are no promoted targets yet, just assume they are empty.
+
+	// With tuf.Targets there is a problem: it sneaks a delegations field into a resulting JSON.
+	// As we don't want that field, a workaround is to use our own simplified type.
+	type TufTargets struct {
+		tuf.SignedCommon
+		Targets tuf.Files `json:"targets"`
+	}
+	targets := TufTargets{}
+	targets.Type = tuf.TUFTypes["targets"]
+	targets.Version += 1
```

Comment:
I am not sure if I understand this `targets.Version += 1`,  it looks like targets.Version will always be equal `1`, how this field `targets.Version` is initialized?

Reply:
You're right, in the initial implementation it'll always be 1 (because this logic only creates the first wave).  My follow-up changes which implement promoted targets should increase a version of promoted targets instead (if present), or fall-back to a current logic if there are no promoted targets yet.